[4.6 Regression] Array allocated with new in a template class is default initialised

 http://gcc.gnu.org/bugzilla/show_bug.cgi?id=54984

--- gcc-4_7-branch/gcc/testsuite/ChangeLog	(revision 192846)
+++ gcc-4_7-branch/gcc/testsuite/ChangeLog	(revision 192847)
@@ -1,6 +1,11 @@
 	PR rtl-optimization/48374
 	* gcc.dg/pr48374.c: New test.
 
+2012-10-26  Paolo Carlini  <paolo.carlini@oracle.com>
+
+	PR c++/54984
+	* g++.dg/template/new11.C: New.
+
 2012-10-22  Andrey Belevantsev  <abel@ispras.ru>
 
 	Backport from mainline
--- gcc-4_7-branch/gcc/testsuite/g++.dg/template/new11.C	(revision 0)
+++ gcc-4_7-branch/gcc/testsuite/g++.dg/template/new11.C	(revision 192847)
@@ -0,0 +1,28 @@
+// PR c++/54984
+// { dg-do run }
+
+int n = 1;
+
+void* operator new(__SIZE_TYPE__)
+{
+  n = -1;
+  return &n;
+}
+
+template <class T>
+struct Foo
+{
+  Foo()
+  : x(new int)
+  {
+    if (*x != -1)
+      __builtin_abort();
+  }
+
+  int* x;
+};
+
+int main()
+{
+  Foo<float> foo;
+}
--- gcc-4_7-branch/gcc/cp/init.c	(revision 192846)
+++ gcc-4_7-branch/gcc/cp/init.c	(revision 192847)
@@ -2794,7 +2794,8 @@
 
       orig_placement = make_tree_vector_copy (*placement);
       orig_nelts = nelts;
-      orig_init = make_tree_vector_copy (*init);
+      if (*init)
+	orig_init = make_tree_vector_copy (*init);
 
       make_args_non_dependent (*placement);
       if (nelts)
--- gcc-4_7-branch/gcc/cp/ChangeLog	(revision 192846)
+++ gcc-4_7-branch/gcc/cp/ChangeLog	(revision 192847)
@@ -1,6 +1,12 @@
 	* class.c (dump_class_hierarchy_r): Cast from pointer via uintptr_t.
 	(dump_vtable): Likewise.
 
+2012-10-26  Paolo Carlini  <paolo.carlini@oracle.com>
+
+	PR c++/54984
+	* init.c (build_new): Don't turn a null *init into a pointer to
+	empty vector orig_init.
+
 2012-09-13  Jason Merrill  <jason@redhat.com>
 
 	PR c++/54511
